075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,368 @@
+3. On the **Nth** reconciliation (rebalancing id done, scaling down can start):
+     * the checks on scaling down the cluster doesn't fail, because the auto-rebalancing was completed and the nodes to be removed don't host any partition replicas anymore.
+     * the `scaleDown()` method can run the scaling down.
+     * the `autoRebalancing()` detects that the `KafkaRebalance` custom resource status reports the `Ready` state (so the rebalancing is completed).
+
+During the above flow, errors can happen before the scaling down:
+
+* There is an error when getting the proposal.
+* After auto-approving the proposal, there is an error when running the rebalance.
+
+In both scenarios, the reconciliation ends with the `KafkaRebalance` custom resource in the `NotReady` state which is reflected into the `Kafka.status.cruiseControl.autoRebalance.state` field.
+
+The reconcile ends but the scale down cannot proceed because of the brokers to be removed are still hosting some partition replicas.
+
+On the next reconciliation, the cluster operator will try to do a scale down again, so running the flow as before.
+The previous `KafkaRebalance` custom resource in the `NotReady` state is overridden to generate a new optimization proposal.
+
+## Affected/not affected projects
+
+This proposal involves the Cluster Operator only.
+More specifically, the development is focused on the `KafkaReconciler` class where methods mentioned in the previous sections need to be added and/or updated.
+Even the `KafkaClusterCreator` should be changed in order to avoid reverting back a scaling down request when an auto-rebalance is configured in the `Kafka` custom resource.
+
+## Compatibility
+
+There are no backward compatibility issues.
+The auto-rebalancing on scaling up and down is optional.
+The user has to define the `spec.cruiseControl.autoRebalance` field on the `Kafka` custom resource in order to have the operator running the rebalancing automatically.
+Without that additional field, the scaling actions are not influenced by any rebalancing operations to take place.
+If the user wants, the rebalancing can be still done with the manual two steps process as today.
+The same way, the new `status.cruiseControl.autoRebalance` field in the `Kafka` custom resource doesn't show up at all if the auto-rebalancing feature is not used.
+
+## Rejected alternatives
+
+### Alternative 1
+
+Like the proposed solution, this alternative is still about leveraging the auto-creation of a `KafkaRebalance` custom resource and the `KafkaRebalanceAssemblyOperator` to take care of it without any further changes.
+The main difference is related to the rebalancing configuration, because it just uses the Cruise Control defaults.
+
+Pros:
+
+* Simpler compared to the proposed solution because no additional `KafkaRebalanceSettings` custom resource is involved.
+  
+Cons:
+
+* No way to customize the configuration (options, goals, ...) for getting a rebalancing proposal.
```

Comment:
I like this idea, but we could simply create a new KafakRebalance annotation "strimzi.io/auto-rebalance" with possible values "add-brokers" and "remove-brokers" to identify auto-rebalance KafakRebalance resources (at most one per type). With that, The only change to Kafka spec would be a boolean field to enable/disable auto rebalance.

Reply:
> What if users supplied there own KafkaRebalance resource and reference it in Kafka resource?

We can go in this direction, but using the `KafkaRebalance` resource as it is can't work. As soon as the user creates it, the operator will start a rebalancing (a manual one, as we support today).

If we really want to avoid a new `KafkaRebalanceSettings` custom resource but using a `KafkaRebalance` instance instead, we would need "something" to differentiate a manual rebalance resource from an auto-rebalance related one. I think we have two options here:

* what @scholzj suggested in another comment by having a `strimzi.io/rebalance: template` annotation on the `KafkaRebalance` which makes clear to the operator, this is a template for a rebalance, don't trigger a rebalancing operation. This "template" rebalance could be referenced within the `Kafka` resource, as well as by any other actual `KafkaRebalance` (used to start a manual rebalancing). So the result is still a settings/template resource but without adding a new custom resource. When scaling, the operator would still create a new `KafkaRebalance` using the settings from the template (but ignoring the `mode` and `brokers` fields because they are implicitly inferenced by the scaling operation).
* not having a template/settings resource but the user has to provide an actual `KafkaRebalance` with `strimzi.io/rebalance: stop` annotation, which would avoid the operator to trigger a manual rebalancing. On scaling, the operator would apply the `refresh` value. Btw, at the end of the auto-rebalancing, the annotation will not be there anymore (so it's different from when the user creates it the first time with `stop`). Anyway it could be used for next scaling operations just applying `refresh` again.

Tbh I would be for having a template/settings and leaving the logic to create an actual `KafkaRebalance` during scaling for auto-rebalancing. This way the templating could be used even for manual rebalancing when the user would like to share the same rebalancing settings across several `KafkaRebalance` instances (as I mentioned in the proposal, if it's fine with others, I think it would deserve another proposal first).

Reply:
Fine, but how do you discriminate between KafkaRebalance templates for the two scaling operations? The suggestion in my previous comment was to use an annotation with the modes as possible values:

```sh
- strimzi.io/auto-rebalance: add-brokers-template|remove-brokers-template
```

That would work as you describe in the template option, but then you can now easily filter the right template for the scaling operation at hand.

Of course there should be only two templates, and we should check that. Or we can simply get the most recent one by type.

Reply:
Why do you need to discriminate. You are referencing a template from the `Kafka` CR where you are specifying the mode (add or remove). So it's in the `Kafka` CR where you are saying "for add-brokers mode use this template <name>", "for remove-brokers mode use this template <name>". It's exactly the same as the current proposal with `KafkaRebalanceSettings` but using a `KafkaRebalance` as template instead.
I don't get why the settings/template custom resource should be aware of which mode it's used.

> Of course there should be only two templates, and we should check that. Or we can simply get the most recent one by type.

Why only two templates? We can make this work even for the manual rebalance where you have your own `KafkaRebalance` refering a template (as this proposal with settings).
You could even use a template to be referenced for auto-rebalancing or manual rebalancing if you need common goals and configuration.

Reply:
In my original suggestion the .spec.cruiseControl.autoRebalance was a simple boolean value to enable disable the feature, and all the logic was based on the annotation, but I think it's better as you propose, as it is explicit configuration.

Reply:
The `strimzi.io/rebalance: template` approach sounds reasonable to me